### PR TITLE
OSDOCS-335: Included draft of ephemeral storage.

### PR DESCRIPTION
**Pull request by huffmanca**
OSDOCS-335: Included draft of ephemeral storage.

This is for OS 4.0.

### CODE REVIEW

**Comment by huffmanca** [comment]
@openshift/team-documentation ,

Could someone peer review this one?

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Module included in the following assemblies:
+//
+// * storage/application-ephemeral-storage.adoc
+
+[id='strategy-{context}']
+= Strategy
+
+. Determine expected container ephemeral usage.
++
+If your administrator enabled the ephemeral storage technology preview,
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: How do I know if my admin did this?

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+//
+// * storage/application-ephemeral-storage.adoc
+
+[id='strategy-{context}']
+= Strategy
+
+. Determine expected container ephemeral usage.
++
+If your administrator enabled the ephemeral storage technology preview, 
+determine expected mean and peak container ephemeral storage usage,
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: Will your user know to determine the storage usage?

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// * storage/application-ephemeral-storage.adoc
+
+[id='strategy-{context}']
+= Strategy
+
+. Determine expected container ephemeral usage.
++
+If your administrator enabled the ephemeral storage technology preview, 
+determine expected mean and peak container ephemeral storage usage, 
+empirically if necessary, for example, by separate extended. Remember to
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: "separate extended" what?

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+[id='strategy-{context}']
+= Strategy
+
+. Determine expected container ephemeral usage.
++
+If your administrator enabled the ephemeral storage technology preview, 
+determine expected mean and peak container ephemeral storage usage, 
+empirically if necessary, for example, by separate extended. Remember to 
+consider all the processes that might potentially run in parallel in the 
+container. For example, does the main application spawn any ancillary
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: s/spawn/create
IBM Style says:
> :no_entry_sign: spawn (a process) v
> Do not use. Use “create.”

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+= Strategy
+
+. Determine expected container ephemeral usage.
++
+If your administrator enabled the ephemeral storage technology preview, 
+determine expected mean and peak container ephemeral storage usage, 
+empirically if necessary, for example, by separate extended. Remember to 
+consider all the processes that might potentially run in parallel in the 
+container. For example, does the main application spawn any ancillary 
+scripts that may require local storage for work files or logging?
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: s/may/

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+. Assess risk for eviction.
++
+Determine the risk appetite for eviction. If the risk appetite is low,
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: I'd define risk appetite. I don't see a style entry for it, but it sounds like borderline jargon that might not translate well.

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+empirically if necessary, for example, by separate extended. Remember to 
+consider all the processes that might potentially run in parallel in the 
+container. For example, does the main application spawn any ancillary 
+scripts that may require local storage for work files or logging?
+
+. Assess risk for eviction.
++
+Determine the risk appetite for eviction. If the risk appetite is low, 
+set the container to request ephemeral storage according to the expected 
+peak usage plus a percentage safety margin. If the risk appetite is
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: I'd s/percentage/

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+. Assess risk for eviction.
++
+Determine the risk appetite for eviction. If the risk appetite is low, 
+set the container to request ephemeral storage according to the expected 
+peak usage plus a percentage safety margin. If the risk appetite is 
+higher, set the container to request ephemeral storage according to the 
+expected mean usage.
+
+. Set container ephemeral storage request.
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: How do I do this?

Comment (kalexand-rh) [reply modules/storage-ephemeral-storage-strategy.adoc]: Also, please rewrite line 25 or line 27 for some variety.

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
++
+Set the container ephemeral storage request based on your risk assessment. 
+The more accurately the request represents the application ephemeral 
+storage usage, the better. If the request is too high, cluster and quota 
+usage will be inefficient. If the request is too low, the chances of 
+application eviction increase.
+
+. Set container ephemeral storage limits, if required.
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: Please rewrite line 33 or line 35 for some variety.

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
++
+Set container ephemeral storage limits, if required. Setting a limit has 
+the effect of immediately stopping a container process if the combined 
+ephemeral storage usage of all processes in the container exceeds the 
+limit. For example, the container might make unanticipated excess 
+ephemeral storage usage obvious early, that is, fail fast, or the 
+container may stop the processes abruptly.
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: s/may/might

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
++
+[IMPORTANT]
+====
+Some {product-title} clusters may require a limit value to be set. Some
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: s/may/might

... How do I know if my cluster requires this? If I'm an admin, how do I set this? I think this note is too vague to be helpful.

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
++
+Set the container ephemeral storage request based on your risk assessment. 
+The more accurately the request represents the application ephemeral 
+storage usage, the better. If the request is too high, cluster and quota 
+usage will be inefficient. If the request is too low, the chances of 
+application eviction increase.
+
+. Set container ephemeral storage limits, if required.
++
+Set container ephemeral storage limits, if required. Setting a limit has
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: How do I know if it's required?

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
++
+[IMPORTANT]
+====
+Some {product-title} clusters may require a limit value to be set. Some 
+may override the request based on the limit, and some application images 
+rely on a limit value being set as this is easier to detect than a 
+request value.
+====
++
+If these limits are set, they should not be set to less than the
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: Rewrite to remove "should."

File: modules/storage-ephemeral-storage-strategy.adoc [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+rely on a limit value being set as this is easier to detect than a 
+request value.
+====
++
+If these limits are set, they should not be set to less than the 
+expected peak container resource usage plus a percentage safety margin.
+
+. Tune the application.
++
+Ensure that the application is tuned with respect to configured request
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-strategy.adoc]: ... How?

File: modules/storage-ephemeral-storage-diagnosing.adoc [hunk 15]
```diff
@@ -0,0 +1,35 @@
+// Module included in the following assemblies:
+//
+// * storage/application-ephemeral-storage.adoc
+
+[id='diagnosing-evicted-pod-{context}']
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-diagnosing.adoc]: This module doesn't belong in this assembly.

File: modules/storage-ephemeral-storage-diagnosing.adoc [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+= Diagnosing an evicted pod
+
+{product-title} might evict a pod from its node when the node’s ephemeral 
+storage is exhausted. Depending on the extent of ephemeral storage 
+exhaustion, the eviction might or might not be graceful. In graceful 
+eviction, the main process, PID 1, of each container receives a 
+SIGTERM signal, then some time later, a SIGKILL signal if the process 
+is still running. In non-graceful eviction, the main process of each 
+container immediately receives a SIGKILL signal.
+
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-diagnosing.adoc]: .Procedure

File: modules/storage-ephemeral-storage-diagnosing.adoc [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+$ oc get pod test -o yaml
+...
+status:
+  message: 'Pod The node was low on resource: [DiskPressure].'
+  phase: Failed
+  reason: Evicted
+----
+
+An evicted pod has phase *Failed* and reason *Evicted*. The evicted pod is
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-diagnosing.adoc]: Needs to be a step.

File: modules/storage-ephemeral-storage-diagnosing.adoc [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+$ oc get pod test -o yaml
+...
+status:
+  message: 'Pod The node was low on resource: [DiskPressure].'
+  phase: Failed
+  reason: Evicted
+----
+
+An evicted pod has phase *Failed* and reason *Evicted*. The evicted pod is 
+not restarted, regardless of the value of *restartPolicy*. However,
```
Comment (kalexand-rh) [inline modules/storage-ephemeral-storage-diagnosing.adoc]: If the replication controller's going to make a new one, why do I need to check for evicted ones? What do I need to do to fix the evicted pods?

**Review by kalexand-rh — COMMENTED** [review]
I think you need to get more data for this assembly.

**Comment by huffmanca** [comment]
After chasing this one done a bit further I've been told that it's intentionally disabled for OS 4.0 and 4.1. I'm closing this PR for now, but will review it again once this functionality has been included.

**Pull request closed without merging**